File: plugins/xcode/Sources/KsealHardenCore/RegistryClient.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// A minimal HTTP request/response pair so the registry client can run against
/// `URLSession` in production and a deterministic mock in tests.
public struct HTTPRequestSpec: Equatable {
    public var url: URL
    public var method: String
    public var headers: [String: String]
    public var body: Data

    public init(url: URL, method: String, headers: [String: String], body: Data) {
        self.url = url
        self.method = method
        self.headers = headers
        self.body = body
    }
}

public struct HTTPResponseSpec {
    public var statusCode: Int
    public var body: Data

    public init(statusCode: Int, body: Data) {
        self.statusCode = statusCode
        self.body = body
    }
}

public protocol HTTPTransport {
    func send(_ request: HTTPRequestSpec) throws -> HTTPResponseSpec
}

/// `URLSession`-backed transport. Synchronous (semaphore) because the tool runs
/// as a one-shot CLI step in CI.
public struct URLSessionTransport: HTTPTransport {
    private let session: URLSession
    private let timeout: TimeInterval

    public init(session: URLSession = .shared, timeout: TimeInterval = 15) {
        self.session = session
        self.timeout = timeout
    }

    public func send(_ request: HTTPRequestSpec) throws -> HTTPResponseSpec {
        var urlRequest = URLRequest(url: request.url, timeoutInterval: timeout)
        urlRequest.httpMethod = request.method
        urlRequest.httpBody = request.body
        for (k, v) in request.headers { urlRequest.setValue(v, forHTTPHeaderField: k) }

        let semaphore = DispatchSemaphore(value: 0)
        var result: Result<HTTPResponseSpec, Error>!
        let task = session.dataTask(with: urlRequest) { data, response, error in
            if let error = error {
                result = .failure(error)
            } else if let http = response as? HTTPURLResponse {
                result = .success(HTTPResponseSpec(statusCode: http.statusCode, body: data ?? Data()))
            } else {
                result = .failure(RegistryError.transport("no HTTP response"))
            }
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()
        return try result.get()
    }
}

/// Configuration for registering a build proof with the control plane.
public struct RegistryConfig {
    public var baseURL: URL
    public var apiKey: String
    public var tenantId: String
    public var appId: String
    public var protectionProfileId: String

    public init(baseURL: URL, apiKey: String, tenantId: String, appId: String, protectionProfileId: String = "") {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.tenantId = tenantId
        self.appId = appId
        self.protectionProfileId = protectionProfileId
    }

    /// Builds a config from environment variables. Returns nil when any required
    /// value is absent, signalling the caller to fall back to offline mode.
    ///
    /// The API key is read from `KSEAL_API_KEY` only — never a flag or a file —
    /// so it is never logged or committed.
    public static func fromEnvironment(_ env: [String: String] = ProcessInfo.processInfo.environment) -> RegistryConfig? {
        guard let base = env["KSEAL_REGISTRY_URL"], let url = URL(string: base),
              let key = env["KSEAL_API_KEY"], !key.isEmpty,
              let tenant = env["KSEAL_TENANT_ID"], !tenant.isEmpty,
              let app = env["KSEAL_APP_ID"], !app.isEmpty
        else { return nil }
        return RegistryConfig(
            baseURL: url,
            apiKey: key,
            tenantId: tenant,
            appId: app,
            protectionProfileId: env["KSEAL_PROTECTION_PROFILE_ID"] ?? ""
        )
    }
}

public enum RegistryError: Error, CustomStringConvertible {
    case transport(String)
    case httpStatus(Int, String)
    case decode(String)

    public var description: String {
        switch self {
        case .transport(let m): return "registry transport error: \(m)"
        case .httpStatus(let code, let body): return "registry returned HTTP \(code): \(body)"
        case .decode(let m): return "registry response decode error: \(m)"
        }
    }
}

/// Where a build proof ended up.
public enum RegistrationOutcome: Equatable {
    /// Registered with the control plane; carries the server build id.
    case registered(buildId: String)
    /// Network/registry unavailable or unconfigured; proof written to `path`.
    case offline(artifactPath: String)
}

/// Client for `RegistryService.CreateBuild` speaking the Connect **JSON**
/// protocol (`POST {base}/kseal.v1.RegistryService/CreateBuild`).
///
/// proto3-JSON rules are honored: field names are camelCase and the int64
/// `versionCode`/`createdAt` are encoded as strings. Auth uses the same bearer
/// API key the control-plane interceptor expects (`Authorization: Bearer …`).
public struct RegistryClient {
    public static let createBuildPath = "/kseal.v1.RegistryService/CreateBuild"

    private let config: RegistryConfig
    private let transport: HTTPTransport

    public init(config: RegistryConfig, transport: HTTPTransport = URLSessionTransport()) {
        self.config = config
        self.transport = transport
    }

    /// Builds the Connect JSON request for a manifest without sending it
    /// (exposed for testing the wire format).
    public func makeCreateBuildRequest(manifest: BuildProofManifest) throws -> HTTPRequestSpec {
        let manifestJSON = try manifest.jsonString()
        let payload: [String: Any] = [
            "tenantId": config.tenantId,
            "appId": config.appId,
            "buildHash": manifest.buildHash,
            "versionName": manifest.versionName,
            // proto3 JSON encodes int64 as a string.
            "versionCode": String(manifest.versionCode),
            "protectionProfileId": config.protectionProfileId.isEmpty ? manifest.protectionProfileId : config.protectionProfileId,
            "manifest": manifestJSON,
        ]
        let body = try JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys])
        // Build the endpoint by string concatenation rather than
        // `appendingPathComponent`, which percent-encodes embedded slashes
        // differently across Foundation implementations (Apple vs. Linux
        // swift-corelibs). The Connect path is a single fixed RPC route.
        let base = config.baseURL.absoluteString
        let trimmedBase = base.hasSuffix("/") ? String(base.dropLast()) : base
        guard let url = URL(string: trimmedBase + Self.createBuildPath) else {
            throw RegistryError.transport("invalid registry URL: \(trimmedBase)\(Self.createBuildPath)")
        }
        return HTTPRequestSpec(
            url: url,
            method: "POST",
            headers: [
                "Content-Type": "application/json",
                "Authorization": "Bearer \(config.apiKey)",
                "Connect-Protocol-Version": "1",
            ],
            body: body
        )
    }

    /// Sends the manifest to the control plane and returns the server build id.
    public func createBuild(manifest: BuildProofManifest) throws -> String {
        let request = try makeCreateBuildRequest(manifest: manifest)
        let response: HTTPResponseSpec
        do {
            response = try transport.send(request)
        } catch {
            throw RegistryError.transport(String(describing: error))
        }
        guard (200..<300).contains(response.statusCode) else {
            throw RegistryError.httpStatus(response.statusCode, String(decoding: response.body, as: UTF8.self))
        }
        guard let obj = try? JSONSerialization.jsonObject(with: response.body) as? [String: Any] else {
            throw RegistryError.decode("response was not a JSON object")
        }
        guard let build = obj["build"] as? [String: Any], let id = build["id"] as? String else {
            throw RegistryError.decode("response missing build.id")
        }
        return id
    }
}

/// Orchestrates registration with a guaranteed offline artifact fallback so a
/// build never fails just because the control plane is unreachable — the proof
/// is durably persisted and can be reconciled later.
public struct BuildProofRegistrar {
    private let transport: HTTPTransport
    private let fileSystem: FileManager

    public init(transport: HTTPTransport = URLSessionTransport(), fileSystem: FileManager = .default) {
        self.transport = transport
        self.fileSystem = fileSystem
    }

    /// Attempts registration; on any failure (or when `config` is nil/`offline`
    /// is set) writes the manifest to `offlineArtifact` and reports `.offline`.
    public func register(
        manifest: BuildProofManifest,
        config: RegistryConfig?,
        offlineArtifact: URL,
        forceOffline: Bool = false
    ) -> Result<RegistrationOutcome, Error> {
        if !forceOffline, let config = config {
            let client = RegistryClient(config: config, transport: transport)
            do {
                let id = try client.createBuild(manifest: manifest)
                return .success(.registered(buildId: id))
            } catch {
                // Fall through to the offline artifact so the build still produces
                // a durable, reconcilable proof.
                return writeOffline(manifest: manifest, to: offlineArtifact)
                    .map { _ in .offline(artifactPath: offlineArtifact.path) }
            }
        }
        return writeOffline(manifest: manifest, to: offlineArtifact)
            .map { _ in .offline(artifactPath: offlineArtifact.path) }
    }

    private func writeOffline(manifest: BuildProofManifest, to url: URL) -> Result<Void, Error> {
        do {
            try fileSystem.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
            try manifest.jsonData().write(to: url, options: .atomic)
            return .success(())
        } catch {
            return .failure(error)
        }
    }
}

File: plugins/gradle/src/main/kotlin/io/kseal/gradle/internal/RegistryClient.kt
package io.kseal.gradle.internal

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * Minimal client for `RegistryService.CreateBuild`, consumed over HTTPS using the
 * Connect unary JSON protocol (a plain `POST <base>/kseal.v1.RegistryService/CreateBuild`
 * with an `application/json` body). The plugin only ever *consumes* this existing
 * API; it never modifies the service.
 *
 * The API key is supplied by the caller (Gradle property / env var) and sent as a
 * Bearer token; it is never logged. `version_code` and `created_at` are int64
 * fields and are encoded/decoded as JSON strings per the protobuf-JSON mapping
 * that connect-go uses.
 */
internal class RegistryClient(
    private val baseUrl: String,
    private val apiKey: String,
    private val httpClient: HttpClient = defaultClient(),
    private val requestTimeout: Duration = Duration.ofSeconds(30),
) {
    data class CreateBuildInput(
        val tenantId: String,
        val appId: String,
        val buildHash: String,
        val versionName: String,
        val versionCode: Long,
        val protectionProfileId: String,
        val manifestJson: String,
    )

    data class CreateBuildResult(
        val buildId: String,
        val httpStatus: Int,
        val rawResponse: String,
    )

    val procedureUrl: String
        get() = baseUrl.trimEnd('/') + PROCEDURE

    fun createBuild(input: CreateBuildInput): CreateBuildResult {
        val body = Json.write(
            linkedMapOf<String, Any?>(
                "tenant_id" to input.tenantId,
                "app_id" to input.appId,
                "build_hash" to input.buildHash,
                "version_name" to input.versionName,
                // int64 -> JSON string (protobuf JSON mapping).
                "version_code" to input.versionCode.toString(),
                "protection_profile_id" to input.protectionProfileId,
                "manifest" to input.manifestJson,
            ),
            indent = false,
        )

        val request = HttpRequest.newBuilder()
            .uri(URI.create(procedureUrl))
            .timeout(requestTimeout)
            .header("Content-Type", "application/json")
            .header("Accept", "application/json")
            .header("Authorization", "Bearer $apiKey")
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .build()

        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        val status = response.statusCode()
        val text = response.body() ?: ""
        if (status !in 200..299) {
            throw RegistryException("CreateBuild failed: HTTP $status ${connectError(text)}")
        }
        val buildId = extractBuildId(text)
        return CreateBuildResult(buildId = buildId, httpStatus = status, rawResponse = text)
    }

    private fun extractBuildId(responseBody: String): String {
        val root = Json.parse(responseBody) as? Map<*, *>
            ?: throw RegistryException("CreateBuild returned a non-object response")
        val build = root["build"] as? Map<*, *>
            ?: throw RegistryException("CreateBuild response missing 'build'")
        return (build["id"] as? String).orEmpty()
    }

    private fun connectError(body: String): String {
        val root = Json.parse(body) as? Map<*, *> ?: return body.take(256)
        val message = root["message"] as? String
        val code = root["code"] as? String
        return listOfNotNull(code?.let { "code=$it" }, message?.let { "message=$it" })
            .joinToString(" ")
            .ifBlank { body.take(256) }
    }

    companion object {
        const val PROCEDURE = "/kseal.v1.RegistryService/CreateBuild"

        private fun defaultClient(): HttpClient =
            HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .followRedirects(HttpClient.Redirect.NEVER)
                .build()
    }
}

internal class RegistryException(message: String) : RuntimeException(message)
